text/0006-monolothic-packaging.md

```diff
@@ -0,0 +1,343 @@
+> (*) This is true for classes, where it will do nominal typing. For interfaces, TypeScript will do structural typing, which means any object that has members matching an interface definition is considered to implement that interface, whether it has been declared to implement it or not, and so a class from one copy of a library may be considered to implement an interface from a different copy.
+Here is an example for a compiler error caused by a dependency mix:
+```
+Unable to compile TypeScript: bin/app.ts(10,36): error TS2345:
+Argument of type 'import("node_modules/@aws-cdk/core/lib/app").App' is not assignable to parameter of type 'import("node_modules/my-module/node_modules/@aws-cdk/core/lib/app").App'. 
+Types have separate declarations of a private property '_assembly'.
+```
+### Implications on CDK
+This situation arises the following use cases:
+* User wants to use an older version than the latest published one (rollback).
+* User is using a 3rd-party construct library.
+Let’s look at these in turn.
+#### User wants to use an old version
+The first one, colloquially called “rollback” is easy to understand and seemingly has a simple solution, so let’s look at it first. This situation comes up when the user is in one of 2 situations:
+* CDK team has just (accidentally) released a buggy version and they want to roll back to an older version.
+* CDK team has just released a breaking change (valid in experimental modules) and the user is not prepared to invest the effort yet to migrate, so they want to stay at an older version.
+For an example, let’s say CDK has just released version *1.13.0* but the user wants to stay at *1.12.0*. The only way they have to achieve this is to control the versions in their own app’s *package.json*, so they write:
+```json
+"dependencies": {
+  "@aws-cdk/aws-ecs": "1.12.0",
+  "@aws-cdk/core": "1.12.0",
+}
+```
+The lack of a caret makes it a fixed version, indicating they really want *1.12.0* and not “at least *1.12.0*” (which would resolve to *1.13.0* in this situation). However, because of the transitive caret dependency the complete dependency graph would look like this:
+```
+app@1.0.0      -> aws-ecs==1.12.0, core==1.12.0
+aws-ecs@1.12.0 -> core^1.12.0
+=== { core@1.13.0 is available } ===> 
+app
++-- core@1.12.0
++-- aws-ecs@1.12.0
+    +-- core@1.13.0
+```
+The dependency tree ends up with 2 versions of *core* in it, and we end up in a broken state.
+THE OBVIOUS SOLUTION to this is to get rid of the transitive caret dependency, and make intra-CDK dependencies fixed version (*==1.12.0, ==1.13.0, etc*).. This puts all control over the specific version that gets pulled in in the hands of the user:
+```
+app@1.0.0      -> aws-ecs==1.12.0, core==1.12.0
+aws-ecs@1.12.0 -> core==1.12.0
+=== { core@1.13.0 is available, but not selected }===> 
+app
++-- core@1.12.0
++-- aws-ecs@1.12.0
+```
+This works fine for the CDK, but breaks in the face of 3rd party construct
```

Comment:
Again, library authors should declare peer dependencies to fix this issue. Monolithic publishing is not the only or even the best way to solve this problem.

Reply:
Yes, the solution is peer dependencies (as mentioned above), but our hyper modularity makes it impractical to apply because consumers of a 3rd party may need to install dozens of transitive dependencies they don’t know or care about and may also be broken when a library adds an additional transitive dependency

Reply:
Yes I understand, but the point I'm making is that you haven't sold me on why just one package as opposed to two or three? Surely 130+ is the _wrong_ number, but why is one the _right_ number?

Reply:
https://en.wikipedia.org/wiki/Zero_one_infinity_rule

Reply:
For real. This is actually the reason

Reply:
This alone does not justify moving to “one” (plus this should be in the doc if this is a primary rationale – not everyone knows about this). My argument is not that we set an arbitrary number. It’s that we continue to allow infinity, but we reset the number of packages we currently have lower. In essence, we didn’t specifically target 130+ packages at the start; we didn’t target a specific number at all. In my proposal we don’t target a number either. We _set_ an initial number as a bound for infinity on the negative end. But we can (and probably will) expand it as AWS expands (towards infinity at this rate, I have no doubt).

Reply:
I will add to the doc. Good call.

**Infinity** does not address the main problem we are trying to solve, which is the fact that consumers of 3rd party library will need to directly depend on all the transitive dependencies of that library.

Reply:
Third party libraries will always need to depend on a transitive dependency, whether it’s just one in your case, or a number less than we have now in mine. The critical issue that I think we’re missing here is that we can solve this issue to the satisfaction of developers and users alike without choosing one. I don’t see the justification for saying “let’s go nuclear - just one dependency!” when many users (myself included) would be more than happy to manage a number say less than ten (to begin with).

Reply:
Let’s say we have 3 modules (A, B and C). consider the case where a 3rd party module peer-depends on module A. 

Now a JavaScript app consumes said 3rd party (defines it under “dependencies”). This means it has to directly also depend on A since A is a peer and won’t be automatically installed.

A month goes by, and our 3rd party evolves in a non-API-breaking way. It now takes a dependency also on B and published a new minor version.

BOOM!

Our app is broken.


The only way I see a solution to this is if we ever only had a single module.

Reply:
That third party app has now violated semver. Even thought the API is non-breaking, it is by definition a breaking change. We cannot not should not be responsible for bad publishing design by third party libs. That could get exhausting very quickly and is not reasonable justification for this entire proposal (eg “people _could_ mess up, so we should bundle the world together forever”)

Reply:
The inconvenience factor of having an additional one or two peer dependencies once in a while, especially when documented in release notes, is not high, nor is it unexpected in the Node world. Dependencies evolve, and people understand this. It’s not the end of the world. The real problem is when there are too many peer dependencies to manage, even with tooling. But the difference between one dep and five deps, for instance, is negligible. Especially when most of them are installed on app init (eg the most common ones)

Reply:
I agree. This is a violation of semver that is caused by how the CDK is packaged. It is unacceptable that 3rd party libraries will need to perform a major version bump every time they add a dependency.

Reply:
This is not inconvenience, it's a design flaw in the CDK that basically means a healthy 3rd party ecosystem cannot be created at the moment.

Reply:
@CaerusKaru I appreciate the discussion, but I think we've exhausted this specific topic. I will try to incorporate some of this feedback under "alternative considered" in the RFC, but otherwise I am resolving this thread for now.